### PR TITLE
chore: Add first logging documentation

### LINKED ISSUE #10907: Add first documentation in repo

### PR DESCRIPTION
**Pull request by timo0**
Fix #10907

### CODE REVIEW

File: platform-sdk/docs/base/logging/logging.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+A particularly useful feature of handlers is the `inheritLevels` property. This boolean property determines whether the handler should inherit the default level configurations set globally. By default, `inheritLevels` is set to `true`. However, it can be turned off to create a handler that focuses exclusively on a specific aspect of logging, such as a particular marker. For example, if you need a handler that only logs entries marked with the `CRYPTO` marker, your configuration would look like this:
+
+```properties
+# Handler specific for CRYPTO marker
+logging.handler.CRYPTO_FILE.level = OFF
+logging.handler.CRYPTO_FILE.marker.CRYPTO = ENABLED
+```
+
+In this configuration, the `CRYPTO_FILE` handler is set to ignore the global log level settings (`level = OFF`) but is specifically enabled to log messages tagged with the `CRYPTO` marker (`marker.CRYPTO = ENABLED`). This setup allows for the creation of dedicated log files or outputs for specific types of log messages, providing a focused view that can be particularly useful in complex systems or during specific types of analysis.
```
Comment (anthony-swirldslabs) [inline platform-sdk/docs/base/logging/logging.md:128]: `the creation of dedicated log files or outputs for specific types of log messages, providing a focused view`

Totally unrelated to this PR, and I'm not sure how exactly this is implemented today, but it seems that this is happening directly in the logger implementation somewhere. So I'm guessing that the code calling the logger might be penalized in terms of the performance to support all these features. It would be nice if these additional mechanisms (unconditional logging based on markers, redirecting some logs to additional destinations, etc.) would be happening in parallel (like on a different thread) or even completely offline to avoid performance degradation associated with logging which might be pretty high because in the end of the day the logging involves I/O.

If that already is the case, then it would be nice to mention this in the document to underscore that all these features don't incur any performance penalties.

File: platform-sdk/docs/base/logging/logging.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+---
+
+
+## Custom Logging Configuration
+
+The configuration syntax is designed for clarity and ease of use. It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.
+For practicality, the `logging.properties` file should be set to reload every second, ensuring up-to-date logging configurations at all times.
+
+Regarding the format of the logging configuration, it should be user-friendly and self-explanatory. Here's an improved representation of the syntax for setting logging levels:
```
Comment (mxtartaglia-sl) [inline platform-sdk/docs/base/logging/logging.md:77]: > Regarding the format of the logging configuration, it should be user-friendly and self-explanatory.

Too wordy... they are design goals...

```suggestion
 Here's a representation of the syntax for setting logging levels:
```

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/docs/base/logging/logging.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+This configuration demonstrates how to set a global default logging level (`INFO`). It also illustrates how to specify logging levels for packages and classes. For example, everything under `com.swirlds.common.crypto` is set to `DEBUG`, except for `com.swirlds.common.crypto.Signature`, which is explicitly set to `WARN`. Similarly, all loggers within `com.hashgraph` default to `WARN`. This approach ensures that loggers inherit the most specific level defined in the configuration, providing both flexibility and precision in logging management.
+
+To provide developers with more control and flexibility in logging, we propose introducing filters for markers. This feature would allow developers to focus on log messages associated with specific markers, regardless of the logger's log level. If there's demand for such functionality among developers, the configuration for marker filters could be structured as follows:
+
+```properties
+# Marker filter configuration
+logging.marker.CONFIG = ENABLED
+logging.marker.CRYPTO = DISABLED
+logging.marker.OTHER = DEFAULT
+```
```
Comment (mxtartaglia-sl) [inline platform-sdk/docs/base/logging/logging.md:98]: I'm not convinced about the style in which it is written...
I think that if the feature is available, it should explain what filters are, how to configure, what they provide after configuring, and leave out phrases like "we propose introducing...", " This feature would allow ..."; "If there's demand for such functionality among developers ..."
And if is not available, I wouldn't make this document the space for discussion.

File: platform-sdk/docs/base/logging/logging.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+4. **Thread Identification**
+5. **Throwable Cause Handling**
+6. **Support for Various Log Levels**
+7. **Marker Support**
+8. **Key-Value Based Metadata**
+9. **Mapped Diagnostic Context (MDC) Support**
+
+---
+
+## Implementing the New Logging API
```
Comment (mxtartaglia-sl) [inline platform-sdk/docs/base/logging/logging.md:24]:
```suggestion
## Using the New Logging API
```

File: platform-sdk/swirlds-logging/src/main/java/com/swirlds/logging/api/internal/level/HandlerLoggingLevelConfig.java [hunk 5]
```diff
@@ -165,7 +165,7 @@ private Map<String, MarkerState> readMarkers(
         configuration.getPropertyNames().filter(n -> n.startsWith(fullPrefix)).forEach(configPropertyName -> {
             final String name = configPropertyName.substring(fullPrefix.length());
             final MarkerState markerState =
-                    configuration.getValue(configPropertyName, MarkerState.class, MarkerState.UNDEFINED);
+                    configuration.getValue(configPropertyName, MarkerState.class, MarkerState.DEFAULT);
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-logging/src/main/java/com/swirlds/logging/api/internal/level/HandlerLoggingLevelConfig.java:168]: Open for consideration: for the sake of maintaining a healthy git history, I think it is worth moving this change to its own ticket; if not, this change would appear in the history as "chore: Add first logging documentation," and we would hide the real motive behind the change.

**Comment by timo0** [comment]
@mxtartaglia-sl @anthony-swirldslabs Thanks for you're feedback! I've revised the documentation please have a second look.

File: platform-sdk/docs/base/logging/logging.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+   ```java
+   import com.swirlds.logging.api.Loggers;
+   ```
+
+2. **Create a Logger Instance**:
+   In your class, declare a `private static final` logger instance. Use the `Loggers.getLogger()` method, passing your class (`MyClass.class`) as a parameter.
+
+   ```java
+   public class MyClass {
+       private static final Logger LOGGER = Loggers.getLogger(MyClass.class);
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/docs/base/logging/logging.md:42]: We will provide a shorthand for IntelliJ or similar but this would introduce much overhead we want to avoid

File: platform-sdk/docs/base/logging/logging.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
+## Hedera Logging API
+
+### Introduction
+The Hedera Logging API is a custom logging facade which is crafted to meet our unique requirements, ensuring compatibility with various software integrations. It acts as a bridge, redirecting events from our facade to SLF4J, providing flexibility for projects dependent on our modules.
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:4]: >  redirecting events from our facade to SLF4J
That is wrong. We do not want to forward anything to SLF4J. SLF4J should forward to our logging system. See diagram at https://github.com/hashgraph/hedera-services/issues/5424

File: platform-sdk/docs/base/logging/logging.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+   ```java
+   import com.swirlds.logging.api.Loggers;
+   ```
+
+2. **Create a Logger Instance**:
+   In your class, declare a `private static final` logger instance. Use the `Loggers.getLogger()` method, passing your class (`MyClass.class`) as a parameter.
+
+   ```java
+   public class MyClass {
+       private static final Logger LOGGER = Loggers.getLogger(MyClass.class);
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:42]: As far as I know we defined to use `logger` instead of `LOGGER`. Pls double check our coding guidelines.

File: platform-sdk/docs/base/logging/logging.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+5. **Throwable Cause Handling**
+6. **Support for Various Log Levels**
+7. **Marker Support**
+8. **Key-Value Based Metadata**
+9. **Mapped Diagnostic Context (MDC) Support**
+
+---
+
+## Using the Hedera Logging API
+
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:25]: Add Warning: DO NOT USE THIS API TODAY! WE STILL WORKING ON THE PERFORMANCE OF THE HANDLERS / OUTPUT.

File: platform-sdk/docs/base/logging/logging.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+# Global default logging level
+logging.level = INFO
+
+# Specific logging levels for packages or classes
+logging.level.com.swirlds.common.crypto = DEBUG
+logging.level.com.swirlds.common.crypto.Signature = WARN
+logging.level.com.hashgraph = WARN
+```
+
+### Levels
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:89]: This headline should be above of the code sample

File: platform-sdk/docs/base/logging/logging.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+To provide developers with more control and flexibility in logging, the configuration supports filters for markers. This feature allows developers to focus on log messages associated with specific markers, regardless of the logger's log level. Here's a sample configuration for marker filters:
+
+```properties
+# Marker filter configuration
+logging.marker.CONFIG = ENABLED
+logging.marker.CRYPTO = DISABLED
+logging.marker.OTHER = DEFAULT
+```
+
+### Markers
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:102]: This headline should be above line 93

File: platform-sdk/docs/base/logging/logging.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+In this configuration, the `logging.marker.NAME` pattern is used, where `NAME` represents the marker's name, and the value can be set to `ENABLED`, `DISABLED`, or `DEFAULT`. For example, `logging.marker.CONFIG = ENABLED` would ensure that all log messages tagged with the `CONFIG` marker are displayed, irrespective of their log level.
+
+A typical logging configuration file, incorporating these marker filters, might look like this:
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:106]: Add a headline like "Using it all together" or so

File: platform-sdk/docs/base/logging/logging.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+A key aspect of the Logging System design is its performance, emphasizing maximum efficiency with as low an impact as possible on the running system.
+
+A particularly useful feature of handlers is the `inheritLevels` property. This boolean property determines whether the handler should inherit the default level configurations set globally. By default, `inheritLevels` is set to `true`. However, it can be turned off to create a handler that focuses exclusively on a specific aspect of logging, such as a particular marker. For example, if you need a handler that only logs entries marked with the `CRYPTO` marker, your configuration would look like this:
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:125]: Can you do one sentence per line? Otherwise the review will become very complexe since a comment can only be done for several sentences.

File: platform-sdk/docs/base/logging/logging.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+```properties
+# Handler specific for CRYPTO marker
+logging.handler.CRYPTO_FILE.level = OFF
+logging.handler.CRYPTO_FILE.marker.CRYPTO = ENABLED
+```
+
+In this configuration, the `CRYPTO_FILE` handler is set to ignore the global log level settings (`level = OFF`) but is specifically enabled to log messages tagged with the `CRYPTO` marker (`marker.CRYPTO = ENABLED`). 
+
+---
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:135]: What is that line for?

File: platform-sdk/docs/base/logging/logging.md [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+
+## Test Support with `WithLoggingMirror`
+
+To improve testing, we've introduced the `WithLoggingMirror` annotation in JUnit 5. This annotation injects a `LoggingMirror` into test methods or classes, providing an isolated environment for logging event analysis. It ensures that tests with this annotation do not run in parallel, maintaining the integrity of each test's logging data.
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:139]: Add link to our doc for junit extensions.

**Review by hendrikebbers — COMMENTED** [review]
Initial review. More comments will be added once we have 1 sentence per line.

**Review by anthony-swirldslabs — DISMISSED** [review]

File: platform-sdk/docs/base/logging/logging.md [hunk 16]
```diff
@@ -0,0 +1,210 @@
+## Hedera Logging API
+
+> [!WARNING]  
+> The API is currently in develpment and is not optimized for production use.
+ It is subject to change without notice.
+
+
+### Introduction
+The Hedera Logging API is a custom logging facade which is crafted to meet our unique requirements, ensuring compatibility with various software integrations.
+It acts as a bridge, redirecting events from eg SLF4J to our facade, providing flexibility for all dependencies used in the project.
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:10]: First of all it acts as our logger that should be used in our code. Second it acts as a bridge. This now sounds like bridge is the main use case.

Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:10]: Add link to SLF4J

File: platform-sdk/docs/base/logging/logging.md [hunk 17]
```diff
@@ -0,0 +1,210 @@
+## Hedera Logging API
+
+> [!WARNING]  
+> The API is currently in develpment and is not optimized for production use.
+ It is subject to change without notice.
+
+
+### Introduction
+The Hedera Logging API is a custom logging facade which is crafted to meet our unique requirements, ensuring compatibility with various software integrations.
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:9]: I would not say that our API is a logging facade. That would mean that we internally use a different logger. Better say Lib or API

File: platform-sdk/docs/base/logging/logging.md [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+It acts as a bridge, redirecting events from eg SLF4J to our facade, providing flexibility for all dependencies used in the project.
+
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:11]: I would add something like: "Our goal is to create a lightweight logging API that perfectly fits to our needs and performance in general"

File: platform-sdk/docs/base/logging/logging.md [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+
+### Architecture and Configuration
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:13]: IO would not mix that 2 points. You have a full section about the config later. So concentrate on Architecture here.

File: platform-sdk/docs/base/logging/logging.md [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+The architecture does not dictate the format of logging output, allowing for a wide range of output options.
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:14]: Before the mentioned points I would add:
as any project of Base the logging is splitted in a public (api) and private (implementation) part and provides extension points next to a very basic functionality. For the logging lib Logging handlers and provider are extension points.
Another goal of the architecture is to not reinvent the wheel how logging in Java works. If you familiar with any Java logging lib you should be more or less ready to use our logger api.

File: platform-sdk/docs/base/logging/logging.md [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+This flexibility permits various configurations, from console outputs to integrations with systems like Graylog or Kibana.
+
+
+### Features of the Logging Facade
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:18]: Again: Not a facade

File: platform-sdk/docs/base/logging/logging.md [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+The facade goes beyond mere message logging, encapsulating rich information within each logging event.
+Key features include:
+
+1. Support for Message Placeholders
+2. Event Origin Tracking
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:23]: limited to the class

File: platform-sdk/docs/base/logging/logging.md [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+3. Timestamp Inclusion
+4. Thread Identification
+5. Throwable Cause Handling
+6. Support for Various Log Levels
+7. Marker Support
+8. Key-Value Based Metadata
+9. Mapped Diagnostic Context (MDC) Support
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:30]: I would say 8+9 are the same.

File: platform-sdk/docs/base/logging/logging.md (hunk 21 above)
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:18]: I would not add the headline but have the content under the architecture headline

File: platform-sdk/docs/base/logging/logging.md [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:31]: Additional points:
- simple property based configuration
- Test Support

File: platform-sdk/docs/base/logging/logging.md [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+## Custom Logging Configuration
+
+The configuration syntax is designed for clarity and ease of use.
+It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.
+
+For practicality, the `logging.properties` file should be set to reload every second, ensuring up-to-date logging configurations at all times.
+
+The format of the logging configuration is user-friendly and self-explanatory.
+Here's an example representation of the syntax for setting logging levels:
```
Comment (hendrikebbers) [inline platform-sdk/docs/base/logging/logging.md:90]: Move that line below the headline

**Review by hendrikebbers — COMMENTED** [review]
Review of first part of the doc.

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/docs/base/logging/logging.md [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+To further refine our logging system, we have incorporated handlers for more granular control over logging behavior. 
+Each handler can be distinctly named and configured using the prefix `logging.handler.NAME`, where `NAME` serves as a unique identifier. 
+This structure allows for the application of handler-specific settings. 
+For instance, `logging.handler.NAME.level` is used to set the logging level for a specific handler, ensuring that all previously discussed features such as marker filters and log level settings are compatible. 
+This setup is instrumental in creating dedicated log files or outputs for specific types of log messages, offering a focused view that is particularly useful in complex systems or during targeted analyses. 
+
+A key aspect of the Logging System design is its performance, emphasizing maximum efficiency with as low an impact as possible on the running system.
+
+
+A particularly useful feature of handlers is the `inheritLevels` property.
```
Comment (jsync-swirlds) [inline platform-sdk/docs/base/logging/logging.md:157]: The example doesn't show using `inheritLevels`.  Do we need a separate example for this?

File: platform-sdk/docs/base/logging/logging.md [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+   ```
+
+   In this example, the `{}` placeholder in the string will be replaced by `"world"`, resulting in the log message `"Hello, world!"`.
+
+## Custom Logging Configuration
+
+The configuration syntax is designed for clarity and ease of use.
+It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.
+
+For practicality, the `logging.properties` file reload every second, ensuring up-to-date logging configurations at all times.
```
Comment (jsync-swirlds) [inline platform-sdk/docs/base/logging/logging.md:91]: nit: `reload` => `will reload`

**Review by jsync-swirlds — DISMISSED** [review]
Looks good, just a couple minor questions.

**Review by mxtartaglia-sl — APPROVED** [review]
